lib/tune.cpp

```diff
@@ -152,9 +152,36 @@ namespace quda
   const map &getTuneCache() { return tunecache; }
 
   /**
-   * Deserialize tunecache from an istream, useful for reading a file or receiving from other nodes.
+   * @brief Distribute the tunecache from a given rank to all other nodes.
+   * @param[in] root_rank From which global rank to do the broadcast
+   * @param[out] tc Where we wish to receive the tunecache.  This
+   * defaults to the local tunecache.
+   */
+  static void broadcastTuneCache(int32_t root_rank = 0, map &tc_recv = tunecache);
+
+  void joinTuneCache(const std::vector<int> &global_tune_rank)
+  {
+    // vector of the split tunecaches
+    std::vector<map> split_tc(global_tune_rank.size());
+    // broadcast each tunecache to every process
+    for (auto i = 0u; i < global_tune_rank.size(); i++) {
+      broadcastTuneCache(global_tune_rank[i], split_tc[i]);
+      if (comm_rank() == global_tune_rank[i]) split_tc[i] = tunecache;
+      logQuda(QUDA_DEBUG_VERBOSE, "i = %d tune_rank = %d tc size = %lu\n", i, global_tune_rank[i], split_tc[i].size());
+    }
+
+    // now merge the maps
+    for (auto i = 0u; i < global_tune_rank.size(); i++) { tunecache.merge(split_tc[i]); }
```

Comment:
This needs to be modified to make the merge deterministic across ranks

Reply:
After digging into the documentaiton on `std::map::merge`, it looks like this may not do what we want. Based on https://en.cppreference.com/w/cpp/container/map/merge , the `merge` routine doesn't do explicit copies: "No elements are copied or moved, only the internal pointers of the container nodes are repointed." This means that if/when `split_tc[i]` falls out of scope, its contents are invalidated, and as such the pointers to objects from `split_tc[i]` now in in `tunecache` are also invalid.

A more concrete code example of my thought:

```
void add_things(std::map<...> &base) {
  std::map<...> local_thing;
  // populate local_thing
  base.merge(local_thing);
  // local_thing falls out of scope
}

...

std::map<...> blah = { ... };
// I want to add things to blah
add_things(blah);

// "blah" now, in theory, contains elements that point to elements of "local_thing", but since "local_thing" fell out of scope those pointers are now invalid
```

Reply:
I've confirmed that `std::map::merge` does do what we want.  When `split_tc[I]` goes out scope there is no issue.  The text regarding `..the internal pointers of the container nodes are repointed...` concern the underlying container of the respective maps.  While a `std::map` doesn't use a (doubly) linked list (will use a red-black tree instead), it's instructive to think in these terms: we splice in various entries by updating the appropriate next and prev pointers of adjacent entries in the list we're inserting an entry into.  No data is moved, only the pointers updated.